src/iio_wrapper.cpp

```diff
@@ -26,24 +26,90 @@
 #include <cstdlib>
 #include <sstream>
 #include <rclcpp/rclcpp.hpp>
+#include <unistd.h>
 
-struct iio_context* IIOWrapper::m_network_context = nullptr;
+struct iio_context* IIOWrapper::m_local_context = nullptr;
+
+struct iio_device* IIOWrapper::m_dev = nullptr;
+struct iio_device* IIOWrapper::m_devtrigger = nullptr;
+struct iio_channel** IIOWrapper::m_channels = nullptr;
+struct iio_buffer* IIOWrapper::m_device_buffer = nullptr;
+int IIOWrapper::m_channel_count = 0;
+
+std::mutex IIOWrapper::m_mutex;
 
 IIOWrapper::IIOWrapper()
 {
-    if(m_network_context == nullptr)
+    if(m_local_context == nullptr)
+    {
+        m_local_context = iio_create_local_context();
+        if(!m_local_context)
+        {
+            throw std::runtime_error("Exception: local context is null");
+        }
+    }
+
+    if(m_dev == nullptr)
     {
-        std::string uri = "ip:127.0.0.1"; // TODO: pass uri from cmd line
+        m_dev = iio_context_find_device(m_local_context, "adis16505");
+        if(!m_dev)
+        {
+            throw std::runtime_error("Exception: device data is null");
+        }
+    }
 
-        m_network_context = iio_create_context_from_uri(uri.c_str());
+    if(m_devtrigger == nullptr)
+    {
+        m_devtrigger = iio_context_find_device(m_local_context, "adis16505-dev0");
+        if(!m_devtrigger)
+        {
+            throw std::runtime_error("Exception: device trigger data is null");
+        }
+    }
 
-        m_object_context =  iio_context_clone(m_network_context);
+    if(m_channels == nullptr)
+    {
+        m_channel_count = 0;
+
+        for (int i = 0; i < iio_device_get_channels_count(m_dev); ++i) {
+            struct iio_channel *chn = iio_device_get_channel(m_dev, i);
+            if (iio_channel_is_scan_element(chn))
+            {
+                m_channel_count++;
+            }
+        }
+
+        m_channels = (iio_channel **)calloc(m_channel_count, sizeof * m_channels );
+
+        for (int i = 0; i < m_channel_count; ++i)
+        {
+            struct iio_channel *chn = iio_device_get_channel(m_dev, i);
+            if (iio_channel_is_scan_element(chn))
+            {
+                m_channels[i] = chn;
+            }
+        }
+
+        // enable first all channels
+        for (int i = 0; i < m_channel_count; ++i)
+        {
+                iio_channel_enable(m_channels[i]);
+        }
     }
-    else
+
+    iio_device_set_trigger(m_dev, m_devtrigger);
+
+    if(m_device_buffer == nullptr)
     {
-        m_object_context =  iio_context_clone(m_network_context);
+        m_device_buffer = iio_device_create_buffer(m_dev, 1, false);
+        if(!m_device_buffer)
+        {
+            throw std::runtime_error("Exception: device buffer is null");
+        }
+        iio_buffer_set_blocking_mode(m_device_buffer, true);
```

Comment:
let's check if this is truly needed, I think by default the buffer is set in blocking mode